src/hardware/mixer.cpp

```diff
@@ -2123,6 +2123,44 @@ static void SDLCALL MIXER_CallBack([[maybe_unused]] void *userdata,
 static void MIXER_Stop([[maybe_unused]] Section *sec)
 {}
 
+// Parse the volume in string form, either in stereo or mono format,
+// and possibly in decibel format, which is prefixed with a 'd'.
+static std::optional<AudioFrame> parse_volume(const std::string &s)
+{
+	auto to_volume = [](const std::string &s) -> std::optional<float> {
+		constexpr auto min_percent = 0.0f;
+		constexpr auto max_percent = 9999.0f;
+		if (const auto p = parse_value(s, min_percent, max_percent); p)
+			return percent_to_ratio(*p);
+
+		constexpr auto min_db = -40.00f;
+		constexpr auto max_db = 39.999f;
+		constexpr auto decibel_prefix = 'd';
+		if (const auto d = parse_prefixed_value(decibel_prefix, s, min_db, max_db); d)
+			return decibel_to_gain(*d);
+
+		return {};
+	};
+	// single volume value
+	auto parts = split(s, ':');
+	if (parts.size() == 1) {
+		if (const auto v = to_volume(parts[0]); v) {
+			AudioFrame volume = {*v, *v};
+			return volume;
+		}
+	}
+	// stereo volume value
+	else if (parts.size() == 2) {
+		const auto l = to_volume(parts[0]);
+		const auto r = to_volume(parts[1]);
+		if (l && r) {
+			AudioFrame volume = {*l, *r};
+			return volume;
+		}
+	}
+	return {};
+}
+
```

Comment:
Very nice improvement, much cleaner and more robust 💯

Reply:
Thanks for the review, @johnnovak !